### DISCUSSION TITLE
Typescript error on enabling persist middleware

### DISCUSSION THREAD

**Discussion by mohanavijay** [pmndrs/zustand#2577] [Q&A] [answered]
I'm encountering a TypeScript error after enabling persist middleware using a slice pattern. 
Any assistance in resolving this would be greatly appreciated.

[typescript playground link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzQGjgZRizQGF0to4BfOAMyghDgCIAvAVwGciA7AE2YBuAFChIsRHD5oAbjAgQANpzxg0UTsG7U6DJmy68+AehDA+fRWgDuGdENHho8JKBDqd9Riw7cM-U3NLGzs0Yzd1B1EebChaDGQ0OAAhTCh8RWBExGE4OAAjNM4ALjgedhBCqBE8jAtUu1KACgBKOABeAD44GQhzESphaNj47IAFdU1uND4G9MzshFyCotLyyvUBobQAD3F4GABPNTgAWUPCYg6UtIystBFhI5PT9iJ5DWuAbWWvgz8AtI5AplMw8DxZOoALo4X7-IzGNQaLQwMFwCbI6azW4LNAwuG+BERKBoiEyaHCKGPWjsHjIGDACA8OAAczQMDmd0SrVKl2wZEwHwAPOc+bgzm8KBo8F8YTc7Fy0N0lnl0DB2FBmU1OOy2l04E0VXkVnYSnAAIwABlhxrgdWxjQNeu6OpgTW17zQzoNSEKptKfmwADo-Z8ANQW6gtFo2qNbYTIJnaUOK65sjk4+6tR67fYoJPwLhoQjQJLtFDkbAii6ezqtJrLIHyJScBu2pFTN3LW3E91B-sYb2G7u2uD9kOZ7njwex21UaOz41G0dlDDuUrMFMfNBgkftuwMjCZVhoZqBr0dbrDle20Nm88T02LucxvfUEctZafloiIA)

Code to reproduce:-
-----------------
```typescript
import { create, StateCreator } from "zustand";
import { devtools, persist } from "zustand/middleware";
import { immer } from "zustand/middleware/immer";

interface BearSlice {
  bears: number;
  addBear: () => void;
}

interface PersistedBearSlice {
  bears: number;
}

export type MyState = BearSlice;

type Mutators = [
  ["zustand/devtools", never],
  ["zustand/persist", PersistedBearSlice],
  ["zustand/immer", never]
];

function getBearSlice(): StateCreator<MyState, Mutators, [], BearSlice> {
  return (set) => ({
    bears: 10,
    addBear: () => set((state) => ({ bears: state.bears + 1 })),
  });
}

const bearSlice = getBearSlice();

export const useStore = create<MyState>()(
  devtools(
    persist(
      immer((...a) => ({
        ...bearSlice(...a),
      })),
      {
        name: "bearStore",
        partialize: (state) => ({
          bears: state.bears,
        }),
      }
    )
  )
);
```

*****
`Argument of type 'WithImmer<Write<WithDevtools<StoreApi<BearSlice>>, StorePersist<BearSlice, unknown>>>' is not assignable to parameter of type 'WithImmer<Write<WithDevtools<StoreApi<BearSlice>>, StorePersist<BearSlice, PersistedBearSlice>>>'.
  Type 'WithImmer<Write<WithDevtools<StoreApi<BearSlice>>, StorePersist<BearSlice, unknown>>>' is not assignable to type 'Omit<Write<WithDevtools<StoreApi<BearSlice>>, StorePersist<BearSlice, PersistedBearSlice>>, "setState">'.
    The types of 'persist.setOptions' are incompatible between these types.
      Type '(options: Partial<PersistOptions<BearSlice, unknown>>) => void' is not assignable to type '(options: Partial<PersistOptions<BearSlice, PersistedBearSlice>>) => void'.
        Types of parameters 'options' and 'options' are incompatible.
          Type 'Partial<PersistOptions<BearSlice, PersistedBearSlice>>' is not assignable to type 'Partial<PersistOptions<BearSlice, unknown>>'.
            Types of property 'storage' are incompatible.
              Type 'PersistStorage<PersistedBearSlice> | undefined' is not assignable to type 'PersistStorage<unknown> | undefined'.ts(2345)`

**Comment by dai-shi** [comment]
If you just need to make types pass, https://tsplay.dev/w68oGm should do.

By the way, recently, I made a new lib and am trying to promote it. Feedback is welcome: https://github.com/zustandjs/zustand-slices

**Reply by mohanavijay** [reply]
Thanks for the quick response, I need to make types pass. I'm facing another typescript issue when passing the store to the constructor.

[playground link for typescript error](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzQGjgWQFcYtc4BlE7AYXS2j0ujQEExg8BVAZzQCEIhAHYATJujgBfOADMoEEHABEAL0LcSopQG4AUKEixEcEWgBuMCBAA23PGDRRuwDVNnzFq9ZpEB6EMAiItZoAO4Y6Dr64NDwSKAgjm5yCspqGhii-oHBYRFovgmOUbrI1hjc3HAsQqAY1oi6cCgQQhpQhMiWUAAUYFDAZqRwdCKt1gCecBrMAFxwAEaYUOJoAJQITVK6krqlra6ZdQ0AvHBCYdW1IPU96mira3p7aAAehvAwEw6Ly6twZx4-EEolWAB4iFQ0GDVmxgBCJpRSAA+PCQ+hOZHIvTAITYKAyDDINBwPh-azAYmNZpLCLceZCQggJZQPTNDBBMkReY9NYA5FwMwQQJ6Xa6N4fOBfH74RFQgGk8mUtDPaUk9HdKpnADaW21XgyWVMFistiUeAuZkcAF0cHqDT5fA4nC4YOa4MIANZCCChIS2+3pR1FKDuy023TW54yYRdYCtOAAczQMC5KwpxN58yRNDo3QRObIGugdjg2ttioi5AzaAFm2a6BghCgQjgPV4MD5JwFPXrzV+dPmAEYAAx2-twDkiNM8rsCjs9dtQudtpC0pzzDLYAB066qAGo4EOpGs1uOT6K9sgDvB19XlQrk6mlZmni93rEWm14Pd-mdUJg2AFlCyK8j0WzGpYNjcOBE7Os4GiwROcAhou27oRgK69lsyFwOhu4vmgPT4Zh54TpIp5kf2fa4UIGCJPMSh3t0aDmjhyFgBEMDAPUwAqGgPJbus-KruxuF7puUIEXSVHkWeYnbBOaxbMpb66EAA)

```typescript
import { create, Mutate, StateCreator, StoreApi, UseBoundStore } from "zustand";
import { devtools, persist } from "zustand/middleware";
import { immer } from "zustand/middleware/immer";

class Animal {
  constructor(private readonly store: bearStore){
  }
}

const animal = new Animal(useStore);


export type bearStore = UseBoundStore<Mutate<StoreApi<MyState>, Mutators>>;
interface BearSlice {
  bears: number;
  addBear: () => void;
}

export type MyState = BearSlice;

type Mutators = [
  ["zustand/devtools", never],
  ["zustand/persist", unknown],
  ["zustand/immer", never]
];

function getBearSlice(): StateCreator<MyState, Mutators, [], BearSlice> {
  return (set) => ({
    bears: 10,
    addBear: () => set((state) => ({ bears: state.bears + 1 })),
  });
}

const bearSlice = getBearSlice();

export const useStore = create<MyState>()(
  devtools(
    persist(
      immer((...a) => ({
        ...bearSlice(...a),
      })),
      {
        name: "bearStore",
        partialize: (state) => ({
          bears: state.bears,
        }),
      }
    )
  )
);
```

***
Argument of type 'UseBoundStore<WithImmer<Write<WithDevtools<StoreApi<BearSlice>>, StorePersist<BearSlice, { bears: number; }>>>>' is not assignable to parameter of type 'bearStore'.
  Type 'UseBoundStore<WithImmer<Write<WithDevtools<StoreApi<BearSlice>>, StorePersist<BearSlice, { bears: number; }>>>>' is not assignable to type 'Omit<Write<WithDevtools<StoreApi<BearSlice>>, StorePersist<BearSlice, unknown>>, "setState">'.
    The types of 'persist.setOptions' are incompatible between these types.
      Type '(options: Partial<PersistOptions<BearSlice, { bears: number; }>>) => void' is not assignable to type '(options: Partial<PersistOptions<BearSlice, unknown>>) => void'.
        Types of parameters 'options' and 'options' are incompatible.
          Type 'Partial<PersistOptions<BearSlice, unknown>>' is not assignable to type 'Partial<PersistOptions<BearSlice, { bears: number; }>>'.
            Types of property 'deserialize' are incompatible.
              Type '((str: string) => StorageValue<unknown> | Promise<StorageValue<unknown>>) | undefined' is not assignable to type '((str: string) => StorageValue<{ bears: number; }> | Promise<StorageValue<{ bears: number; }>>) | undefined'.ts(2345)

**Reply by dai-shi** [reply, marked as answer]
Here you go: https://tsplay.dev/wRb31w

**Reply by mohanavijay** [reply]
Thank you so much!

**Reply by mohanavijay** [reply]
> If you just need to make types pass, https://tsplay.dev/w68oGm should do.
> 
> By the way, recently, I made a new lib and am trying to promote it. Feedback is welcome: https://github.com/zustandjs/zustand-slices

I will definitely check out the new library. It looks intriguing!
